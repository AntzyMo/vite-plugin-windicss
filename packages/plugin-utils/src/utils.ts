import { toArray } from '@antfu/utils'

export type { Arrayable } from '@antfu/utils'
export { toArray, partition, slash } from '@antfu/utils'

export type NestedArrayable<T> = T | (T | T[])[]

export function flattenArray<T>(v: NestedArrayable<T>): T[] {
  return toArray(v).flat() as T[]
}

export function mergeArrays<T>(...args: (NestedArrayable<T> | undefined)[]): T[] {
  return args.flatMap(i => flattenArray(i || []))
}

export function kebabCase(str: string) {
  return str.replace(/([a-z])([A-Z])/g, '$1-$2').toLowerCase()
}

export function include<T>(set: Set<T>, v: T[] | Set<T>) {
  for (const i of v)
    set.add(i)
}

export function exclude<T>(set: Set<T>, v: T[] | Set<T>) {
  for (const i of v)
    set.delete(i)
}

export function escapeRegExp(str: string) {
  return str.replace(/[.*+?^${}()|[\]\\]/g, '\\$&') // $& means the whole matched string
}
