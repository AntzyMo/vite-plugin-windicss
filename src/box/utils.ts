import { regexClassGroup } from './constants'

export function toArray<T>(v: T | T[]): T[] {
  if (Array.isArray(v))
    return v
  return [v]
}

export function kebabCase(str: string) {
  return str.replace(/([a-z])([A-Z])/g, '$1-$2').toLowerCase()
}

export function include<T>(set: Set<T>, v: T[] | Set<T>) {
  for (const i of v)
    set.add(i)
}

export function exclude<T>(set: Set<T>, v: T[] | Set<T>) {
  for (const i of v)
    set.delete(i)
}

export function transfromGroups(str: string) {
  return str.replace(
    regexClassGroup,
    (_, a: string, b: string) => b.split(/\s/g).map(i => `${a}:${i}`).join(' '),
  )
}
