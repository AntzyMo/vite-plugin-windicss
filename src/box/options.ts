/* eslint-disable no-use-before-define */

import type { Config as WindiCssOptions } from 'windicss/types/interfaces'
import { defaultAlias, TagNames } from './constants'
import { kebabCase, toArray } from './utils'

export { WindiCssOptions }

export interface WindiBoxOptions {
  /**
   * Options for windicss/tailwindcss.
   * Also accepts string as config file path.
   *
   * @default 'tailwind.config.js'
   */
  windicssOptions?: WindiCssOptions | string

  /**
   * Safe class names to be always included.
   */
  safelist?: string | string[]

  /**
   * Enabled windicss preflight (a.k.a TailwindCSS style reset)
   *
   * @default true
   */
  preflight?: boolean | {
    /**
     * Safelist to always included
     */
    safelist?: string | string[]

    /**
      * Alias for resolving preflight
      */
    alias?: Record<string, TagNames>

    /**
     * @default true
     */
    includeBase?: boolean

    /**
     * @default true
     */
    includeGlobal?: boolean

    /**
     * @default true
     */
    includePlugin?: boolean
  }

  /**
   * Scan the files and extract the usage
   *
   * @default true
   */
  scan?: boolean | {
    /**
     * Auto scan on startup
     *
     * @default true
     */
    runOnStartup?: boolean

    /**
     * Directories to search for classnames
     *
     * @default 'src'
     */
    dirs?: string[]

    /**
     * File extension to search for classnames
     *
     * @default 'html', 'vue'
     */
    fileExtensions?: string[]

    /**
     * Exclude globs
     *
     * @default []
     */
    exclude?: string[]

    /**
     * Include globs
     *
     * @default []
     */
    include?: string[]
  }

  /**
   * Transform CSS for `@apply` directive
   *
   * @default true
   */
  transformCSS?: boolean

  /**
   * Sort the genrate utilities
   *
   * @default true
   */
  sortUtilities?: boolean

  /**
   * Name for debug
   *
   * @default 'windi-box'
   * @internal
   */
  name?: string

  /**
   * CWD
   *
   * @default process.cwd
   * @internal
   */
  root?: string
}

export function resolveOptions(options: WindiBoxOptions) {
  const {
    windicssOptions = 'tailwind.config.js',
    scan = true,
    preflight = true,
    transformCSS = true,
    sortUtilities = true,
    root = process.cwd(),
  } = options

  const preflightOptions = Object.assign(
    {
      includeBase: true,
      includeGlobal: true,
      includePlugin: true,
      alias: {},
    },
    typeof preflight === 'boolean' ? {} : preflight,
  )

  const scanOptions = Object.assign(
    {
      fileExtensions: ['html', 'vue', 'md', 'pug', 'jsx', 'tsx', 'svelte'],
      dirs: ['src'],
      exclude: [] as string[],
      include: [] as string[],
      runOnStartup: true,
    },
    typeof scan === 'boolean' ? {} : scan,
  )

  const safelist = new Set(toArray(options.safelist || []).flatMap(i => i.split(' ')))

  preflightOptions.alias = Object.fromEntries(
    Object.entries({
      ...defaultAlias,
      ...preflightOptions.alias,
    }).filter(([k, v]) => [kebabCase(k), v]),
  )

  return {
    ...options,
    windicssOptions,
    scan: Boolean(scan),
    scanOptions,
    preflight: Boolean(preflight),
    preflightOptions,
    transformCSS,
    sortUtilities,
    safelist,
    root,
  }
}

export type ResolvedOptions = ReturnType<typeof resolveOptions>
