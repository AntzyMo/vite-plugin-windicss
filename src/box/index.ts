import { promises as fs, existsSync } from 'fs'
import { join, resolve } from 'path'
import WindiCssProcessor from 'windicss'
import { StyleSheet } from 'windicss/utils/style'
import { CSSParser } from 'windicss/utils/parser'
import fg from 'fast-glob'
import _debug from 'debug'
import { regexQuotedString, regexClassSplitter, regexClassCheck, regexHtmlTag, preflightTags, htmlTags, defaultAlias, TagNames, regexClassGroup } from './constants'
import { resolveOptions, WindiCssOptions, WindiBoxOptions } from './options'

import { toArray, kebabCase, include, exclude } from './utils'

export type WindiBox = ReturnType<typeof createBox>

export { preflightTags, htmlTags, defaultAlias }
export type { WindiBoxOptions, TagNames }

export function createBox(_options: WindiBoxOptions = {}) {
  const options = resolveOptions(_options)

  const {
    name,
    config,
    scan: enabledScan,
    scanOptions,
    transformCSS: enableCssTransform,
    preflight: enablePreflight,
    preflightOptions,
    sortUtilities,
    root,
    safelist,
  } = options

  const debug = {
    config: _debug(`${name}:config`),
    debug: _debug(`${name}:debug`),
    compile: _debug(`${name}:compile`),
    glob: _debug(`${name}:glob`),
    detect: _debug(`${name}:detect`),
  }

  let processor: WindiCssProcessor
  let configFilePath: string | undefined

  const regexId = new RegExp(`\\.(?:${scanOptions.fileExtensions.join('|')})$`, 'i')

  const configSafelist = new Set<string>()

  const classesGenerated = new Set<string>()
  const classesPending = new Set<string>()
  const tagsGenerated = new Set<string>()
  const tagsPending = new Set<string>()
  const tagsAvailable = new Set<string>()

  function loadConfiguration() {
    let resolved: WindiCssOptions = {}
    if (typeof config === 'string') {
      const path = resolve(root, config)
      if (!existsSync(path)) {
        console.warn(`[${name}] config file "${config}" not found, ignored`)
      }
      else {
        try {
          delete require.cache[require.resolve(path)]
          // eslint-disable-next-line @typescript-eslint/no-var-requires
          resolved = require(path)
          configFilePath = path
          configSafelist.clear()
          // @ts-expect-error
          include(configSafelist, resolved?.purge?.options?.safelist || resolved?.purge?.options?.whitelist || [])
        }
        catch (e) {
          console.error(`[${name}] failed to load config "${config}"`)
          console.error(`[${name}] ${e.toString()}`)
          process.exit(1)
        }
      }
    }
    else {
      resolved = config
    }

    debug.config(JSON.stringify(resolved, null, 2))
    return resolved
  }

  function initWindicss() {
    return new WindiCssProcessor(loadConfiguration())
  }

  let _searching: Promise<void> | null

  async function scan() {
    if (!_searching) {
      const { fileExtensions, dirs, include, exclude } = scanOptions
      _searching = (async() => {
        const globs = dirs.map(i => join(i, `**/*.{${fileExtensions.join(',')}}`).replace(/\\/g, '/'))
        globs.unshift('index.html')
        globs.unshift(...include)

        debug.glob('globs', globs)

        const files = await fg(
          globs,
          {
            cwd: root,
            ignore: ['node_modules', '.git', ...exclude],
            onlyFiles: true,
            absolute: true,
          },
        )

        files.sort()

        debug.glob('files', files)

        const contents = await Promise.all(
          files
            .filter(id => isDetectTarget(id))
            .map(id => fs.readFile(id, 'utf-8')),
        )

        for (const content of contents)
          extractFile(content)
      })()
    }

    return _searching
  }

  function isDetectTarget(id: string) {
    return id.match(regexId)
  }

  function extractFile(code: string) {
    // classes
    Array.from(code.matchAll(regexQuotedString))
      .flatMap(m => (m[2] || '')
        .replace(regexClassGroup, (v) => {
          if (!classesGenerated.has(v))
            classesPending.add(v)
          return ''
        })
        .split(regexClassSplitter),
      )
      .filter(i => i.match(regexClassCheck))
      .forEach((i) => {
        if (!i || classesGenerated.has(i))
          return
        classesPending.add(i)
      })

    if (enablePreflight) {
      // preflight
      Array.from(code.matchAll(regexHtmlTag))
        .flatMap(([, i]) => i)
        .forEach((i) => {
          if (!tagsAvailable.has(i))
            i = preflightOptions.alias[kebabCase(i)]
          if (!tagsAvailable.has(i))
            return
          tagsPending.add(i)
          tagsAvailable.delete(i)
        })
    }

    debug.detect('classes', classesPending)
    debug.detect('tags', tagsPending)
  }

  function transformCSS(css: string) {
    if (!enableCssTransform)
      return css
    const style = new CSSParser(css, processor).parse()
    return style.build()
  }

  let style: StyleSheet = new StyleSheet()
  let _cssCache: string | undefined

  async function generateCSS() {
    if (enabledScan && scanOptions.runOnStartup)
      await scan()

    let changed = false

    if (classesPending.size) {
      const result = processor.interpret(Array.from(classesPending).join(' '))
      if (result.success.length) {
        include(classesGenerated, result.success)
        classesPending.clear()
        debug.compile(`compiled ${result.success.length} classes`)
        debug.compile(result.success)

        style = style.extend(result.styleSheet)
        changed = true
      }
    }

    if (enablePreflight && tagsPending.size) {
      const preflightStyle = processor.preflight(
        Array.from(tagsPending).map(i => `<${i}`).join(' '),
        preflightOptions.includeBase,
        preflightOptions.includeGlobal,
        preflightOptions.includePlugin,
      )
      style = style.extend(preflightStyle, true)
      include(tagsGenerated, tagsPending)
      tagsPending.clear()
      changed = true
    }

    if (changed || !_cssCache) {
      if (sortUtilities)
        style.sort()

      _cssCache = style.build()
    }
    return _cssCache
  }

  function clearCache() {
    style = new StyleSheet()
    _cssCache = undefined

    const preflightSafelist = toArray(preflightOptions?.safelist || []).flatMap(i => i.split(' '))

    include(classesPending, configSafelist)
    include(classesPending, safelist)
    include(classesPending, classesGenerated)

    include(tagsPending, tagsGenerated)
    include(tagsPending, preflightTags)
    include(tagsPending, preflightSafelist)
    include(tagsAvailable, htmlTags as any as string[])

    exclude(tagsAvailable, preflightTags)
    exclude(tagsAvailable, preflightSafelist)

    classesGenerated.clear()
    tagsGenerated.clear()
  }

  function init() {
    processor = initWindicss()
    clearCache()
  }

  return {
    options,
    init,
    extractFile,
    generateCSS,
    clearCache,
    transformCSS,
    isDetectTarget,
    scan,

    classesGenerated,
    classesPending,
    tagsGenerated,
    tagsPending,
    tagsAvailable,

    get configFilePath() {
      return configFilePath
    },
  }
}
