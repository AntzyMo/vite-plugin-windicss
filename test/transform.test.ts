import { transfromGroups } from '../src/box/utils'

describe('transfrom', () => {
  it('group basic', async() => {
    expect(
      transfromGroups('bg-white font-light sm:hover:(bg-gray-100 font-medium)'),
    ).toMatchSnapshot()
  })
})
