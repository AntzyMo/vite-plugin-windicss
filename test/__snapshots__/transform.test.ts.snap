// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`transfrom group basic 1`] = `"bg-white font-light sm:hover:bg-gray-100 sm:hover:font-medium"`;
